turbopack/crates/turbo-persistence/src/compaction/interval_map.rs

```diff
@@ -0,0 +1,166 @@
+    {
+        match self
+            .intervals
+            .binary_search_by_key(&location, |point| point.start)
+        {
+            Ok(index) => index,
+            Err(index) => {
+                // If the location does not exist, we need to insert a new interval
+                let value = if index > 0 {
+                    self.intervals[index - 1].value.clone()
+                } else {
+                    None
+                };
+                self.intervals.insert(
+                    index,
+                    IntervalPoint {
+                        start: location,
+                        value,
+                    },
+                );
+                index
+            }
+        }
+    }
+
+    /// Applies the update function to all values in the specified range.
+    pub fn update(&mut self, range: &(u64, u64), mut update: impl FnMut(&mut T))
+    where
+        T: Default + Clone,
+    {
+        let start = range.0;
+        let end = range.1;
+        if start > end {
+            return;
+        }
+
+        let start = self.ensure_point(start);
+        if end == u64::MAX {
+            for i in start..self.intervals.len() {
+                update(self.intervals[i].value.get_or_insert_default());
+            }
+        } else {
+            let end = self.ensure_point(end + 1);
+
+            for i in start..end {
+                update(self.intervals[i].value.get_or_insert_default());
+            }
+        }
+    }
+
+    /// Tests if any values in the specified range satisfy the predicate.
+    pub fn test(&self, range: &(u64, u64), mut predicate: impl FnMut(&T) -> bool) -> bool {
+        let start = range.0;
+        let end = range.1;
+        if start > end {
+            return false;
+        }
+
+        let start_index = self
+            .intervals
+            .binary_search_by_key(&start, |point| point.start)
+            .unwrap_or_else(|x| x);
+
+        let end_index = match self
+            .intervals
+            .binary_search_by_key(&end, |point| point.start)
+        {
+            Ok(index) => index + 1,
+            Err(index) => index,
+        };
+
+        for i in start_index..end_index {
+            if let Some(value) = &self.intervals[i].value
+                && predicate(value)
+            {
+                return true;
+            }
+        }
+        false
+    }
+
+    /// Returns an iterator over the ranges and their associated values.
+    pub fn ranges(&self) -> impl Iterator<Item = ((u64, u64), &T)> {
+        (0..self.intervals.len()).filter_map(move |i| {
+            let start = self.intervals[i].start;
+            let end = if i + 1 < self.intervals.len() {
+                self.intervals[i + 1].start - 1
+            } else {
+                u64::MAX
+            };
+            self.intervals[i]
+                .value
+                .as_ref()
+                .map(|value| ((start, end), value))
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I've written a range/interval set (not a map though, so that code wouldn't be useful here) before in Rust for a side project. I had a ton of trouble testing it.

What I ended up doing was creating a new nested loops to exhaustively test all possible set of ranges with N elements and with a custom `TinyInt` type that had a small range from 0...M, where N and M are small (e.g. N < 4, M < 6).